source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -1,27 +1,62 @@
 
-LocalRateLimiterImpl::LocalRateLimiterImpl(const std::chrono::milliseconds fill_interval,
-                                           const uint32_t max_tokens,
-                                           const uint32_t tokens_per_fill,
-                                           Event::Dispatcher& dispatcher)
-    : fill_interval_(fill_interval), max_tokens_(max_tokens), tokens_per_fill_(tokens_per_fill),
-      fill_timer_(fill_interval_ > std::chrono::milliseconds(0)
+LocalRateLimiterImpl::LocalRateLimiterImpl(
+    const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
+    const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher,
+    const Protobuf::RepeatedPtrField<
+        envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>& descriptors)
+    : fill_timer_(fill_interval > std::chrono::milliseconds(0)
                       ? dispatcher.createTimer([this] { onFillTimer(); })
-                      : nullptr) {
-  if (fill_timer_ && fill_interval_ < std::chrono::milliseconds(50)) {
+                      : nullptr),
+      time_source_(dispatcher.timeSource()) {
+  if (fill_timer_ && fill_interval < std::chrono::milliseconds(50)) {
     throw EnvoyException("local rate limit token bucket fill timer must be >= 50ms");
   }
 
-  tokens_ = max_tokens;
+  token_bucket_.max_tokens_ = max_tokens;
+  token_bucket_.tokens_per_fill_ = tokens_per_fill;
+  token_bucket_.fill_interval_ = absl::FromChrono(fill_interval);
+  tokens_.tokens_ = max_tokens;
 
   if (fill_timer_) {
-    fill_timer_->enableTimer(fill_interval_);
+    fill_timer_->enableTimer(fill_interval);
+  }
+
+  for (const auto& descriptor : descriptors) {
+    RateLimit::LocalDescriptor new_descriptor;
+    for (const auto& entry : descriptor.entries()) {
+      new_descriptor.entries_.push_back({entry.key(), entry.value()});
+    }
+    RateLimit::TokenBucket token_bucket;
+    token_bucket.fill_interval_ =
+        absl::Milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(descriptor.token_bucket(), fill_interval, 0));
+    if (token_bucket.fill_interval_ % token_bucket_.fill_interval_ != absl::ZeroDuration()) {
+      throw EnvoyException(
+          "local rate descriptor limit is not a multiple of token bucket fill timer");
+    }
+    token_bucket.max_tokens_ = descriptor.token_bucket().max_tokens();
+    token_bucket.tokens_per_fill_ =
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(descriptor.token_bucket(), tokens_per_fill, 1);
+    new_descriptor.token_bucket_ = token_bucket;
+
+    auto token_state = std::make_unique<RateLimit::TokenState>();
+    token_state->tokens_ = token_bucket.max_tokens_;
+    token_state->fill_time_ = time_source_.monotonicTime();
+    new_descriptor.token_state_ = token_state.get();
+    descriptor_tokens_.emplace_back(std::move(token_state));
+
+    auto result = descriptors_.emplace(new_descriptor);
+    if (!result.second) {
+      throw EnvoyException("duplicate descriptor in the local rate descriptor");
```

Comment:
nit: can you print something about the descriptor like key/value/etc.?

Reply:
Ping on this

Reply:
Done.